### PR TITLE
Add WebhookTypes and EventForType methods

### LINKED ISSUE #2863: Support enumerating the known event types
It looks like `go-github` has [an internal list of all the GitHub event types ](https://github.com/google/go-github/blob/899235e0a9d750d6fecf9048a676046d50f9d4a3/github/messages.go#L46).  I'd like to be able to iterate over the list in my own code where I route events based on event type.  I'm thinking of exposing the following methods:

```golang
// WebhookTypes returns a list of all the known GitHub event type strings supported by go-github.
func WebhookTypes() []string {
  // ...
}

// EventForType returns an empty struct matching the specified GitHub event type.
func EventForType(messageType string) any {

}
```

The second method would extract a lot of `Event.ParsePayload` https://github.com/google/go-github/blob/master/github/event.go#L30, so it wouldn't be much additional code.

### PR DESCRIPTION
**Pull request by evankanderson**
Fixes #2863

### CODE REVIEW

File: github/messages.go
Comment (gmlewis) [inline github/messages.go:333]: This thing is such a pain to remember for new PR contributors.

What if we were to update `eventTypeMapping` to include both fields we want to look up?

For example:

```go
type eventTypeTuple struct {
  eventName string
  emptyStruct interface{}
}

var (
  eventTypeMapping = map[string]eventTypeTuple{
    "branch_protection_rule": {"BranchProtectionRuleEvent", &BranchProtectionRuleEvent{}},
    "check_run":              {"CheckRunEvent", &CheckRunEvent{}},
    ...
  }
)
```

Then we could change this `EventForType` to also use the same map.
Finally, every time a new event is added, only one line needs updating in the `eventTypeMapping`.

Thoughts?

**Review by gmlewis — CHANGES_REQUESTED** [review]

**Review by gmlewis — CHANGES_REQUESTED** [review]
Could you please also add tests for requesting a `nil` event, an unknown event, and any other place where we pass in garbage to make sure it behaves properly in all cases?

File: github/messages.go
Comment (evankanderson) [reply github/messages.go:333]: I thought about this -- we need to make a _copy_ of the object, which means using reflection given that the declared type is `interface{}`.  If the performance cost seems acceptable, that sounds great to me.

File: github/event.go
Comment (evankanderson) [reply github/event.go:33]: Sure.  I got thrown off because the other methods were in `github-accessors.go`, so I missed that they existed.

File: github/event.go
Comment (gmlewis) [inline github/event.go:34]: Can this use an accessory?

Comment (evankanderson) [reply github/event.go:34]: Yeah... it turns out this is also subtly wrong if payload is `nil`, I think.  I'll have a fix shortly, along with tests for `ParsePayload` on an `Event{}` (zero-valued struct).

File: github/messages.go
Comment (gmlewis) [inline github/messages.go:48]: Hey - here's a thought to avoid reusing the same pointer over and over again... what if these lines all returned a struct by _VALUE_ instead of by _reference_?

```suggestion
		"branch_protection_rule":         BranchProtectionRuleEvent{},
```

And then each call to the function would return a brand new pointer.

Thoughts?

Comment (evankanderson) [reply github/messages.go:48]: Let me make sure we don't end up aliasing the returned values.

Comment (evankanderson) [reply github/messages.go:48]: Ah, I could do that, but would it be any better than the current `reflect.New` usage?

File: github/event_test.go
Comment (gmlewis) [inline github/event_test.go:36]: Also, please:

```suggestion
  var nilEvent *Event
  nilEvent.ParsePayload()
}
```

File: github/messages.go
Comment (gmlewis) [inline github/messages.go:343]: Let's please return the happy path at the end, which is idiomatic Go style:

```suggestion
	if prototype == nil {
		return nil
	}
	// return a _copy_ of the pointed-to-object.  Unfortunately, for this
	// we need to use reflection.
	return reflect.New(reflect.TypeOf(prototype).Elem()).Interface()
```

File: github/messages.go
Comment (gmlewis) [reply github/messages.go:48]: Oh, I gotcha.  No, what you have looks good, thanks!

File: github/event_test.go
Comment (evankanderson) [reply github/event_test.go:36]: Done.

Comment (evankanderson) [reply github/event_test.go:43]: (I also checked that both of these returned error, which seems correct.)

File: github/messages.go
Comment (evankanderson) [reply github/messages.go]: Done.  I also referenced this issue so people could find [the comment above](https://github.com/google/go-github/pull/2865#discussion_r1290433599).

File: github/messages.go
Comment (evankanderson) [reply github/messages.go:48]: Oh, one more thought over the weekend: except for "GitHubAppAuthorizationEvent", all of these follow a regular snake_case --> TitleCase conversion.  I _could_ build this map with a type list and reflection, plus a special case for "GitHub" --> "github", but that seemed _too_ clever, so I didn't.

Comment (gmlewis) [reply github/messages.go:48]: Yes, we prefer simplicity over cleverness, so thank you for keeping it simpler... it is greatly appreciated.  😄

File: github/messages.go
Comment (gmlewis) [inline github/messages.go:345]: Although I got a chuckle out of `weevils`, I'm thinking that we might want to be a bit more friendly to our friends where English is their Nth language (for N > 1), so let's please revert this to "evils".  Thank you. 😄

**Review by gmlewis — CHANGES_REQUESTED** [review]
Thank you, @evankanderson !
A few more tweaks, please, then I think we should be ready for a second LGTM+Approval from any other contributor to this repo before merging.

File: github/messages.go
Comment (evankanderson) [reply github/messages.go:325]: I really wish that go would include https://pkg.go.dev/k8s.io/apimachinery/pkg/util/sets#Set, but in the meantime, I can sort the list.

File: github/messages.go
Comment (evankanderson) [reply github/messages.go:345]: Yeah, I had to try.  That's what code review is for!

File: github/event.go
Comment (evankanderson) [reply github/event.go:39]: Ah -- this subtly changes the semantics to always return `nil` (rather than an empty struct) if the `json.Unmarshal` fails.  I'm okay with that.

Comment (evankanderson) [reply github/event.go:39]: And if I'm doing that, there's no point in naming the return parameters, so I won't.

**Comment by evankanderson** [comment]
Ready for another look / another reviewer, I think.

File: github/event.go
Comment (gmlewis) [inline github/event.go:33]:
```suggestion
	payload := EventForType(typeToMessageMapping[e.GetType()])
```

Comment (evankanderson) [reply github/event.go]: Done.  Earlier, I'd had some trouble with the compiler getting the types correct, but I think that was with the map-of-structs rather than map-of-pointers-to-structs solution.

**Comment by evankanderson** [comment]
Do you want me to flatten my commits?

**Comment by gmlewis** [comment]
> Do you want me to flatten my commits?

No need to flatten any commits, as we always "squash and merge" in this repo.  Thank you, though.

**Review by gmlewis — APPROVED** [review]
@evankanderson - this looks great to me!
Thank you for your patience and your willingness to experiment with this one.
I think this looks super clean and will help with reducing not only the maintenance burden of this repo, but also make it _much_ easier for contributors to add new event types as GitHub keeps cranking them out.  :grin:

LGTM.

Awaiting second LGTM+Approval from any other contributor to this repo before merging.

**Comment by gmlewis** [comment]
@googlebot - why have you been silent on this PR? What happened to your automated CLA checks?!?

**Comment by gmlewis** [comment]
Ah, is @googlebot dead and been replaced by @google-cla instead?

**Comment by gmlewis** [comment]
@google-ospo-team - can you please investigate why this PR was never checked by the @google-cla bot?
I can't move forward on this PR until it verifies the CLA is signed by all parties.
Thank you!

**Comment by gmlewis** [comment]
CLA has been cleared.  Thank you, @google-ospo-team !

Now awaiting second LGTM+Approval from any other contributor to this repo before merging.

**Comment by gmlewis** [comment]
@Parker77 - do you have time for a code review, please?

File: github/messages.go
Comment (Parker77) [inline github/messages.go:114]:
```suggestion
	// forward mapping of event types to the string names of the structs
```

**Review by Parker77 — APPROVED** [review]
LGTM!

**Comment by gmlewis** [comment]
Thank you, @Parker77 !
Merging.

**Pull request merged**
